arangod/Aql/HashedCollectExecutor.cpp

```diff
@@ -0,0 +1,299 @@
+      _producingOutput(false),
+      _isInitialized(false),
+      _aggregatorFactories() {
+  _aggregatorFactories = createAggregatorFactories(_infos);
+};
+
+HashedCollectExecutor::~HashedCollectExecutor() {
+  // Generally, _allGroups should be empty when the block is destroyed - except
+  // when an exception is thrown during getOrSkipSome, in which case the
+  // AqlValue ownership hasn't been transferred.
+  destroyAllGroupsAqlValues();
+}
+
+void HashedCollectExecutor::destroyAllGroupsAqlValues() {
+  for (auto& it : _allGroups) {
+    for (auto& it2 : it.first) {
+      const_cast<AqlValue*>(&it2)->destroy();
+    }
+  }
+}
+
+void HashedCollectExecutor::consumeInputRow(InputAqlItemRow& input) {
+  TRI_ASSERT(input.isInitialized());
+
+  decltype(_allGroups)::iterator currentGroupIt;
+  bool newGroup;
+  std::tie(currentGroupIt, newGroup) = findOrEmplaceGroup(input);
+
+  // reduce the aggregates
+  AggregateValuesType* aggregateValues = currentGroupIt->second.get();
+
+  if (_infos.getAggregateVariables().empty()) {
+    // no aggregate registers. simply increase the counter
+    if (_infos.getCount()) {
+      // TODO get rid of this special case if possible
+      TRI_ASSERT(!aggregateValues->empty());
+      aggregateValues->back()->reduce(EmptyValue);
+    }
+  } else {
+    // apply the aggregators for the group
+    TRI_ASSERT(aggregateValues->size() == _infos.getAggregatedRegisters().size());
+    size_t j = 0;
+    for (auto const& r : _infos.getAggregatedRegisters()) {
+      if (r.second == ExecutionNode::MaxRegisterId) {
+        (*aggregateValues)[j]->reduce(EmptyValue);
+      } else {
+        (*aggregateValues)[j]->reduce(input.getValue(r.second));
+      }
+      ++j;
+    }
+  }
+}
+
+void HashedCollectExecutor::writeCurrentGroupToOutput(OutputAqlItemRow& output) {
+  // build the result
+  TRI_ASSERT(!_infos.getCount() || _infos.getCollectRegister() != ExecutionNode::MaxRegisterId);
+
+  auto& keys = _currentGroup->first;
+  TRI_ASSERT(_currentGroup->second != nullptr);
+
+  TRI_ASSERT(keys.size() == _infos.getGroupRegisters().size());
+  size_t i = 0;
+  for (auto& key : keys) {
+    output.cloneValueInto(_infos.getGroupRegisters()[i++].first,
+                          _lastInitializedInputRow, key);
+    const_cast<AqlValue*>(&key)->erase();  // to prevent double-freeing later
+  }
+
+  if (!_infos.getCount()) {
+    TRI_ASSERT(_currentGroup->second->size() == _infos.getAggregatedRegisters().size());
+    size_t j = 0;
+    for (auto const& r : *(_currentGroup->second)) {
+      output.cloneValueInto(_infos.getAggregatedRegisters()[j++].first,
+                            _lastInitializedInputRow, r->stealValue());
+    }
+  } else {
+    // set group count in result register
+    TRI_ASSERT(!_currentGroup->second->empty());
+    output.cloneValueInto(_infos.getCollectRegister(), _lastInitializedInputRow,
+                          _currentGroup->second->back()->stealValue());
+  }
```

Comment:
@hkernbach In this whole method, do we really have to clone any values? We emit each group once and then never need it again, so I think stealing/moving the values would be prudent, so we avoid copying and can possibly free memory earlier.